### PR TITLE
KAFKA-3103; Transient Failure in testIsrAfterBrokerShutDownAndJoinsBack

### PR DESCRIPTION
**Pull request by rowdyrabbit**
All three defects have the same root cause. 

Root cause is ClientUtils.fetchTopicMetadata returns the BrokerEndPoints in a non-deterministic order, so we need to sort the expected endpoints and the received endpoints so we can correctly compare them.

### CODE REVIEW

File: core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala [hunk 1]
```diff
@@ -201,7 +201,7 @@ abstract class BaseTopicMetadataTest extends ZooKeeperTestHarness {
                                 2000, 0)
         metadata.topicsMetadata.nonEmpty &&
           metadata.topicsMetadata.head.partitionsMetadata.nonEmpty &&
-          expectedIsr == metadata.topicsMetadata.head.partitionsMetadata.head.isr
+          expectedIsr.sorted == metadata.topicsMetadata.head.partitionsMetadata.head.isr.sorted
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: Why not simply use `sortBy(_.id)`? It doesn't seem that we need ordering outside of tests, so I think I'd prefer that we don't introduce the `Ordered` interface in `BrokerEndPoint` until that is the case.

Comment (ijuma) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: Also, `Ordering` is generally preferred over `Ordered` in Scala.

**Comment by ijuma** [comment]
Nice catch, thanks for fixing this! I left a minor comment in the PR. Also, can you please edit the PR title to be something like:

`KAFKA-3103; Transient Failure in testIsrAfterBrokerShutDownAndJoinsBack` ?

I marked the other tickets as duplicates, so we just need to refer to the one actually.

File: core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala (hunk 1 above)
Comment (rowdyrabbit) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: No problem, I'll make the change and re-submit the PR, thanks for reviewing!

Comment (ijuma) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: Great. You can add an additional commit to this PR?

Comment (rowdyrabbit) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: I was going to back out the last commit and re-submit with a new modified commit, is that ok?

File: core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala [hunk 2]
```diff
@@ -215,7 +215,7 @@ abstract class BaseTopicMetadataTest extends ZooKeeperTestHarness {
 
   @Test
   def testIsrAfterBrokerShutDownAndJoinsBack {
-    val numBrokers = 2 //just 2 brokers are enough for the test
+    val numBrokers = 4 //just 2 brokers are enough for the test
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: Is this change intentional?

File: core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala (hunk 1 above)
Comment (ijuma) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: Sure.

File: core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala (hunk 2 above)
Comment (rowdyrabbit) [inline core/src/test/scala/unit/kafka/integration/BaseTopicMetadataTest.scala]: No, I'm fixing it up now.

**Comment by ijuma** [comment]
LGTM

**Comment by granthenke** [comment]
LGTM

**Comment by ijuma** [comment]
I've reopened KAFKA-3155 as @guozhangwang pointed out that the stacktrace looks different. @rowdyrabbit, can you explain how it's the same issue?

**Comment by rowdyrabbit** [comment]
Actually, it looks like KAFKA-3155 **is** a separate issue, I think it's just named poorly - it should be called:

Transient Failure in kafka.api.BaseProducerSendTest.testFlush 

and one occurrence is in kafka.api.PlaintextProducerSendTest.testFlush

**Comment by ijuma** [comment]
Thanks @rowdyrabbit

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk and 0.9.0.
